### ISSUE TITLE
Resource Pack Sound Loss (boss fights)

### ISSUE THREAD

**Issue by O13SC3N3** [OreCruncher/BetterRain#63]
Dynamic Surroundings; 1.0.4.5BETA
Minecraft: 1.7.10
Forge: 1566 and forge 1558

With the recent update Resource Packs that change in game music (Haven't tested other sounds like Mobs) is lost. This includes End, Credits, Boss, Nether 1-4, Creative 1-6, Calm 1-3, Hal 1-4, Nuance 1-2, Piano 1-3.

I have tried increasing the sound channels up to 16, 32, 48, 64 with no favorable result.

1.0.4.3BETA has the same effect, which I overlooked.
And all versions down to version 1.0.3.1BETA.

I have 1.0.3.1 BETA in my current release of my private server pack, so it wasn't noticed until today. I was putting the final touches on my updates, script changes etc. And noticed I didn't feel all alone in a dark evil nether, even tho it has dust storms now...

@OreCruncher I checked each version and somewhere after 1.0.3.1BETA its lost, I can supply you a download link to the sound pack if you want to test it yourself.

**Comment by OreCruncher** [comment]
A link to the sound pack would be appreciated.

EDIT: Question on the loss.  Is it that these sounds don't play, or they start playing, then stop?

**Comment by O13SC3N3** [comment]
Its that they are completely ignored, Vanilla sounds/music will play. Resource Pack Override is blocked somehow.

**Comment by OreCruncher** [comment]
That is really weird.  The mod doesn't involve itself with sound registration process and the like.  These sounds don't even trigger with /playsound?

**Comment by O13SC3N3** [comment]
Naw, (its being uploaded send link as soon as its done)... Yeah I don't know what the deal is I went to check on all the versions after 1.0.31 and somehow it stopped...

**Comment by OreCruncher** [comment]
RE: Potion Hud - does that black box clipping show when you hit "E" and look at the potion list to the left of the dialog?

**Comment by OreCruncher** [comment]
Took a look inside the resource pack - didn't see a sounds.json file.  I am a n00b when it comes to resource packs but I think I would have expected to see that file.  (It tells Minecraft/Forge how to integrate the sound information into the runtime environment.)

**Comment by O13SC3N3** [comment]
(Potion Hud) - http://i.imgur.com/USSyOas.png, maybe conflict with another mod? Cuz I deleted the config file, and all I did was lowwerd the hud to not cover the clock.

**Comment by O13SC3N3** [comment]
Never had to use it. Only resource packs that had a sounds json file was useally for specific events. I can give you a link to the modpack if you want, Ive used this resource pack for the last year tho :p

**Comment by OreCruncher** [comment]
PotionHud - must be something with how my overlay sets up it's canvas.  I will have to dig deeper.  (Can you write up a separate issue so that it can be tracked?)

Sounds - I assume that if you remove Dynamic Surroundings the sounds come back?

**Comment by O13SC3N3** [comment]
Exactly remove Dynamic Surroundings, everything works fine. Revert all the way back to 1.0.31 everything works fine.

**Comment by OreCruncher** [comment]
For the sounds - is there something I can do to reliably trigger a sound from the pack?

**Comment by O13SC3N3** [comment]
No idea, All the sound pack does it replace Vanillas Music. Also replaces Cave, thunder, rain sounds.

**Comment by OreCruncher** [comment]
Yeah - figured out the thunder one.  I triggered a thunderstorm using /weather.  Right now it seems to be working for me on the surface in a rain storm.  I get your thunder and rain sounds as well as ambient music.  I have a private build I am working on that changes how the mod tracks sounds.  Let me package it up and make it available to you for testing on your end to see if it address.

EDIT:  1.7.10 right?

EDIT2: https://github.com/OreCruncher/BetterRain/releases/tag/v1.7.10-1.0.4.5BETAHF

EDIT3: This release also contains weapon "swooshes" when using a ItemSword or ItemAxe.  I mention it in case you hear any sounds.

**Comment by O13SC3N3** [comment]
Do you get a new beat when you face the ender dragon? If so then it works for you.

**Comment by OreCruncher** [comment]
I didn't hear it.  Of course I had my biome sound playing and it could have hidden it.  But I did hear the tracks in Overworld.

**Comment by O13SC3N3** [comment]
Nope, sounds are lost in the end, nether. Overworld ill have to test more.

Side Note: The weapon swoosh is that for all contact with entities can it be singled out ie adding nasty sounds of axes chopping into flesh zombies, cows etc, bones skeleton?

edit gonna try something, didnt work, tried disabling biome sounds see if it did. Nether is default minecraft music.

Its so weird how the mod nukes a resource packs sounds looks like mainly in other dimensions. And yet music via vanilla is free to play. Im at a loss.

**Comment by OreCruncher** [comment]
Yeah, this sound thing has me baffled.  As mentioned prior Dynamic Surroundings doesn't mess with the sound manager - it just plays the sound.  Could be a bug in the sound resource loading - will have to trace it and see.

Right now the weapons swoosh is tied to the AttackEntityEvent which occurs when a player makes contact with a mob.  I filter based on the equipped item and use that as the basis for deciding the sound to play.  The logic doesn't consider the mob target.  I will have to consider adding that, and provide config options to allow an author to further refine which sound to play - similar to what I do for biomes with condition tokens.

**Comment by OreCruncher** [comment]
Just went to the nether and your background music is playing.  Beginning to think it may be something else in the pack - not sure what.  You have the pack online somewhere?

EDIT: Get your background music in the end, too.

**Comment by O13SC3N3** [comment]
Hmm Im not sure, you change any settings? Cuz i've been testing in 2 environments, One with just Forge 1566, Dynamic Surroundings, and the Resource Pack. The other with the mod pack. Both seem to only allow the Overworld music now. Nether and End doesn't work.

And the pack I can only distribute privately due to FTB Terms.

**Comment by O13SC3N3** [comment]
seems like end plays the standard Overworld music now, Just not triggering the Boss Fight event or even the fact you are in the End. The end has 3 tracks, Boss is actually shared depending on entity. 

resource pack/assets/minecraft/sounds/music/game/end

boss.ogg
end.ogg
credits.ogg

resource pack/assets/minecraft/sounds/music/game/nether

nether1.ogg
nether2.ogg
nether3.ogg
nether4.ogg

resource pack/assets/minecraft/sounds/music/game

the main folder holds the main songs, there is another folder for creative tracks.

outside of that folder sounds/music/menu is for the title menu.

Its like the resource pack with dynamic surroundings can not go further into the pack to pull from. This confuses me lol.

**Comment by OreCruncher** [comment]
Anything showing up in the client log related to resource loading?

(I am using Forge 1614 in case somehow it makes a difference.)

**Comment by O13SC3N3** [comment]
Nothing that sticks out. Heres a copy of the log http://pastebin.com/raw/jBgyivnb

**Comment by OreCruncher** [comment]
Shot in the dark - the only resource I have in the Minecraft namespace is the particles.png.  Open the Dynamic Surroundings JAR and remove assets.minecraft.*.  Maybe it is somehow interfering with the stuff Rev5 has (noticed that it replaces particles as well).

**Comment by O13SC3N3** [comment]
kk

**Comment by O13SC3N3** [comment]
Whats changed from 1.0.31 to current? I just don't see how anything you added will limit a resource pack since then.

Tested it with a few other Sound Packs out there. They all suffer the same loss, Nether, End, Boss.

EDIT _Limited Down_; So after testing my pack and others like Snos and such. I decided to Work Backwards and Forwards for each update. 1.0.3.1 and 1.0.3.2 Wanted CodeChickenCore. cool, Thought that was it. Upgraded to 1.0.4.0 removed CodeChickenCore to check. Everything worked. Moved up to 1.0.4.1 Resource Pack Loss of Nether and the End. So from 1.0.4.0 to 4.1 is the change.

**Comment by OreCruncher** [comment]
Try changing "Fancy Cloud Handling" to false.  Let me know if it addresses.

**Comment by O13SC3N3** [comment]
Hmm no, at least not with BETA5 anyways, checking others.

**Comment by OreCruncher** [comment]
OK - here is what I am seeing.  Using MultiMC to create an instance, and the only mod is Dynamic Surroundings, and of course the resource pack you provided.  I disabled my sounds for The End so I can hear what is going on.  [I also enabled Minecraft's debug trace](http://www.minecraftforum.net/forums/support/unmodified-minecraft-client/tutorials-and-faqs/2134680-tut-advanced-debugging-minecraft) so I could see when it fired off a sound and when it thought it was removed.

Loaded up a creative instance and went to The End.  I clobbered the dragon and jumped through the portal back.  Got the credits.  The log trace did not show any attempt to play the credit/end music.

I removed Dynamic Surroundings and loaded up the world again.  Went to The End, and jumped back through the gate to get the credits.  Still no trace or attempt to play the sound.

Honestly don't know what is going on.  I did get your sounds while in The End so I think everything should be integrated properly.  The log trace shows the reload of the sound information and no errors were reported.

EDIT: This was with Forge 1558

**Comment by OreCruncher** [comment]
Went to the end again and the End sound is now playing:

[21:09:30] [Client thread/DEBUG]: Playing sound minecraft:sounds/music/game/end/end.ogg for event minecraft:music.game.end as channel 9370513d-205d-4c9f-8db2-8c0decd15545

**Comment by O13SC3N3** [comment]
there should be a boss fight track when you combat the dragon. edit wrong link here

https://drive.google.com/file/d/0B3xshvaIB3CDTjhqTXZfU0pBSlk/view?usp=sharing

**Comment by OreCruncher** [comment]
Yep.  Wasn't one.  And the log showed that it didn't try to play it.  Do you have any other mods that are trying to overwrite parts of the Minecraft assets?  You know Dynamic Surroundings has that particle sheet and I am wondering if another mod could be doing the same things you are.  Try naming your pack something other than Rev5 - maybe ZRev5.  Forge may put the packs in alpha order to process them.

**Comment by O13SC3N3** [comment]
Thing is when I go back to Dynamic Surroundings 3.2 and 4.0  it works fine

**Comment by OreCruncher** [comment]
Yep - understand.

**Comment by OreCruncher** [comment]
lol - just went to the end again looking for a boss fight and guess what:

[21:31:28] [Client thread/DEBUG]: Playing sound minecraft:sounds/music/game/end/boss.ogg for event minecraft:music.game.end.dragon as channel 67af7747-59e6-4a6a-b7a8-216cd0283103

I have Dynamic Surroundings loaded, too.

**Comment by O13SC3N3** [comment]
yeah its only working for me with versions 4.0 and below. Dont understand why its not for the others.

**Comment by OreCruncher** [comment]
I am using a *.0.4.6BETA I am working on.  I have seen what you experienced, and the log shows that the sounds aren't being played.  Currently looking into the triggers for the sounds - may be something in there that can give clues.

**Comment by O13SC3N3** [comment]
yeah something since 4.1 and up stops it. Not sure from what you list in your changelogs that would effect that.

**Comment by OreCruncher** [comment]
You try out changing that fancy cloud rendering to false?  Just found something in the Minecraft code that could screw up boss music.  For the life of me I don't know why they coded it that way.

For fancy cloud rendering I actually have a WorldProvider wrapper that I insert bewteen the World object and the real WorldProvider object.  This allows me to handle requests for things like fog color and cloud height so I can do some magic.  What I am seeing in the boss music code is that they are checking the provider class type to see if it is the provider for Nether and/or the End rather than checking the dimension ID.  Since my shim is in there its not going to pass those tests, ergo no fancy boss/end/enther music.

As indicated above the boss music did play eventually.  However, in my case I have the fancy cloud rendering turned OFF so it would pass these "tests" that Vanilla is doing.

**Comment by O13SC3N3** [comment]
So how does that play against your mod?

**Comment by OreCruncher** [comment]
Not understanding the question.  Based on what I see in the Minecraft code what Dynamic Surroundings is doing for fancy cloud support breaks the boss music for the nether (Wither) and The End (dragon).  I will need to come up with a different way of handling in order to keep Minecraft happy.

I should also mention this to the guy that does StellarSky - he does the same thing for handling "real world" sun/moon behaviors for Minecraft.

**Comment by O13SC3N3** [comment]
yeah changing the cloud fancy worked, it didnt register in the modpack the first time I tried. I deleted the config, closed the game out a second time. And now it works.

So with that being disabled what I ment was what do I lose with Dynamic Surroundings? And why is it effected only after version 3.2/4.0 Beta 4.1 and up?

**Comment by OreCruncher** [comment]
By turning it off you lose the sky and clouds darkening during rain.  And the reason it showed up in those BETAs was when I implemented the shim.  Prior I used ASM to modify the java code on the fly, and that should only be the last resort.  Currently testing a fix.

**Comment by O13SC3N3** [comment]
Ooh, you know I was going to ask you about that... I never saw the sky or clouds darken.. Maybe it was the resource pack? You know you have the right to say "Dynamic Surroundings" is NOT COMPAITABLE with Boss, Nether, End Music Resource Packs :p. But I do appreciate you spending basically all day with me trying to figure this out lol.

**Comment by OreCruncher** [comment]
No worries.  My goal is to make the mod as compatible as possible.  Though I do provide out of the box content and stuff, the real purpose of the mod is to allow mod pack authors to customize their packs.

As for the sky/clouds darkening it is dependent on the rain fall strength.  Set the time to day, do a /toggledownfall, and then do a /rain 100

**Comment by O13SC3N3** [comment]
well assuming if fancy cloud is enabled that is. I actually been testing your mod against a lot of "fancy" mods, so far it held up beside this issue today. Optifine/FastCraft/FastFPS/Better Foliage/Dynamic Lighting/ Custom Sky texture packs. I tinker a lot. I have a tendency to find issues...

**Comment by OreCruncher** [comment]
That's cool that you do.  There are so many mods and combinations of things it is hard to say something is bullet proof.  The Minecraft modding is so dynamic it is a constant break/find/fix cycle. :)  Eventually things will stabilize.

BTW, anything dealing with shaders will more than likely interfere with rain, fog, and aurora rendering.  I know SEUS messes with it. Doesn't break the mod - just that stuff doesn't render.

**Comment by O13SC3N3** [comment]
Yeah Ive been experimenting. Ive hosted a private server for the last 2 years with private packs. this year I decided to do all the politics to get hosted through ftb etc. So I useally spend more time adding, removing, scripting, then I do playing with the few friends/family that play it. Im almost at the point now, I can sit down and play lol. But half the time I fix something, get a new update. I gotta change something else.

here soon I gotta cover it with an "Application" according to FTB's third party pack rules, to open the server up to a few more players.

**Comment by OreCruncher** [comment]
Yeah - getting a modpack through the gauntlet can be a pain.  It's one of the reasons I make my mods MIT licensed - I don't want to become a hurdle.  And I am familiar with the pack work time vs. play time.  Lucky for me I ran servers for a small group of folks that were mature and avoided creating messes. :)

**Comment by OreCruncher** [comment]
BETA build pushed to CurseForge: http://minecraft.curseforge.com/projects/dynamic-surroundings/files
